HPCC-22690 Change JlibSemTestStress::testSimple to use realistic timings and prevent to getting stuck indefinitely.

AttilaVamos opened the pull request:
Change upper time limit from '2 * time' to 'time + <one minute>'

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [X] Send notifications about my Pull Request position in Smoketest queue.
- [X] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Process of PR-12973, label: HPCC-22690-fix-7.6.x starts now.
The reason of this test is: Did not build yet.
Commit ID: a5a102c8a95a4ced7a7794183a3377dfe833f78b
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: a5a102c8a95a4ced7a7794183a3377dfe833f78b
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |860 |860 |0 |181 sec  (00:03:01) |
|test (thor) |784 |784 |0 |686 sec  (00:11:26) |
|test (roxie) |935 |935 |0 |242 sec  (00:04:02) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   37 sec (00:00:37) |  258 sec (00:04:18) |    0 sec (00:00:00) |    3 sec (00:00:03) |   19 sec (00:00:19) | 1445 sec (00:24:05) |   22 sec (00:00:22) | 1784 sec (00:29:44) |

AttilaVamos commented:
@jakesmith  Is this what we are talked earlier? (It is finished on my Ubuntu VM without any error, Earlier it is failed 20 times from 20 runs)

jakesmith commented on testing/unittests/jlibtests.cpp:
```diff
@@ -32,6 +32,8 @@
 
 #include "unittests.hpp"
 
+#define ONE_MINUTE 60000 // msec
```
minor: nicer as a typed static const e.g.
```
static const unsigned oneMinute = 60000; // msec
```

jakesmith commented on testing/unittests/jlibtests.cpp:
```diff
@@ -129,7 +131,7 @@ class JlibSemTestStress : public CppUnit::TestFixture
             CPPUNIT_FAIL(errMsg.str());
         }
-        testTimedElapsed(sem, 5, 1000);
+        testTimedElapsed(sem, 20, 1000);
```
not sure why 5 change to 20 here?
(and NB: not in other places, like line 78)

jakesmith commented:
@AttilaVamos - a couple of comments, but yes think a constant max time change is good.

There's also a similar potential problem with testTimedAvailable though.
i.e. it could take >=5 ms.

I think I'd change it to:

```
    void testTimedAvailable(Semaphore & sem)
    {
        //Shouldn't cause a reschedule, definitely shouldn't wait for 100s
        if (!sem.wait(100))
            ASSERT(false);
    }
```

since that's what it's actually testing afaics.

AttilaVamos commented:
@jakesmith updated, please re-check.

HPCCSmoketest commented:
Process of PR-12973, label: HPCC-22690-fix-7.6.x starts now.
The reason of this test is: New commit.
Commit ID: 5017971a315e196424375d6b35d5f062dcde8b92
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 5017971a315e196424375d6b35d5f062dcde8b92
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |860 |860 |0 |210 sec  (00:03:30) |
|test (thor) |784 |784 |0 |713 sec  (00:11:53) |
|test (roxie) |935 |935 |0 |230 sec  (00:03:50) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   40 sec (00:00:40) |  260 sec (00:04:20) |    0 sec (00:00:00) |    2 sec (00:00:02) |   20 sec (00:00:20) | 1487 sec (00:24:47) |   22 sec (00:00:22) | 1831 sec (00:30:31) |

jakesmith commented on testing/unittests/jlibtests.cpp:
```diff
@@ -43,11 +45,9 @@ class JlibSemTest : public CppUnit::TestFixture
         //Shouldn't cause a reschedule, definitely shouldn't wait for 100s
-        ASSERT(taken < 5);
+        if(!sem.wait(100))
+            ASSERT(false)
```
missing ';' (seems to compile anyway, but for consistency)

AttilaVamos replied:
Changed

jakesmith commented:
@richardkchapman - Not sure important, but I think could be targeted to 7.6.0 ?

@AttilaVamos - 1 minor comment.

HPCCSmoketest commented:
Process of PR-12973, label: HPCC-22690-fix-7.6.x starts now.
The reason of this test is: New commit.
Commit ID: 3b04196e5183b3ee73a0535a61412bbe9b46e700
Estimated completion time is ~0.5 hour(s)

AttilaVamos commented:
@jakesmith updated.

jakesmith commented:
@AttilaVamos - looks good, please squash - and I think can be 7.6.0 too.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 3b04196e5183b3ee73a0535a61412bbe9b46e700
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |6 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |27 sec  (00:00:27) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |18 sec  (00:00:18) |
|test (hthor) |860 |860 |0 |174 sec  (00:02:54) |
|test (thor) |784 |784 |0 |708 sec  (00:11:48) |
|test (roxie) |935 |935 |0 |239 sec  (00:03:59) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   30 sec (00:00:30) |  253 sec (00:04:13) |    0 sec (00:00:00) |    3 sec (00:00:03) |   21 sec (00:00:21) | 1431 sec (00:23:51) |   21 sec (00:00:21) | 1759 sec (00:29:19) |

AttilaVamos committed:
HPCC-22690 Change JlibSemTestStress::testSimple to use realistic timings and prevent to getting stuck indefinitely.

Change upper time limit from '2 * time' to 'time + <one minute>'

Tested manually

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

AttilaVamos commented:
@jakesmith squashed and re-targeted.

HPCCSmoketest commented:
Process of PR-12973, label: HPCC-22690-fix-7.6.x starts now.
The reason of this test is: New commit.
Commit ID: 3af84d5f98acbaf1984b913bfeaccc387bdf3ea8
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 3af84d5f98acbaf1984b913bfeaccc387bdf3ea8
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |3 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |27 sec  (00:00:27) |
|setup (thor) |11 |11 |0 |46 sec  (00:00:46) |
|setup (roxie) |11 |11 |0 |26 sec  (00:00:26) |
|test (hthor) |860 |860 |0 |180 sec  (00:03:00) |
|test (thor) |784 |784 |0 |717 sec  (00:11:57) |
|test (roxie) |935 |935 |0 |256 sec  (00:04:16) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   34 sec (00:00:34) |  262 sec (00:04:22) |    0 sec (00:00:00) |    3 sec (00:00:03) |   19 sec (00:00:19) | 1499 sec (00:24:59) |   21 sec (00:00:21) | 1838 sec (00:30:38) |

HPCCSmoketest commented:
Process of PR-12973, label: HPCC-22690-fix-7.6.x starts now.
The reason of this test is: Base branch changed.
Commit ID: 3af84d5f98acbaf1984b913bfeaccc387bdf3ea8
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 3af84d5f98acbaf1984b913bfeaccc387bdf3ea8
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |46 sec  (00:00:46) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |860 |860 |0 |177 sec  (00:02:57) |
|test (thor) |784 |784 |0 |713 sec  (00:11:53) |
|test (roxie) |935 |935 |0 |419 sec  (00:06:59) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   30 sec (00:00:30) |  259 sec (00:04:19) |    0 sec (00:00:00) |    3 sec (00:00:03) |   19 sec (00:00:19) | 1626 sec (00:27:06) |   22 sec (00:00:22) | 1959 sec (00:32:39) |

jakesmith commented:
Looks good.

@richardkchapman - ready to merge
